lib/bundle.js

```diff
@@ -290,13 +299,78 @@ function handleLocalCollisions(trace, initialMainKeys) {
    * @param {string} commonPathFromData - The common path in the file's paths
    * @param {Array} allData -  array of { path, content} objects
    * @param {object} globalReferences - The accumulated global references from all nodes
+   * @param {function} remoteRefResolver - The function that would be called to fetch remote ref contents
    * @returns {object} - The references in current node and the new content from the node
    */
-function getReferences (currentNode, isOutOfRoot, pathSolver, parentFilename, version, rootMainKeys,
-  commonPathFromData, allData, globalReferences) {
+async function getReferences (currentNode, isOutOfRoot, pathSolver, parentFilename, version, rootMainKeys,
+  commonPathFromData, allData, globalReferences, remoteRefResolver) {
   let referencesInNode = [],
     nodeReferenceDirectory = {},
-    mainKeys = {};
+    mainKeys = {},
+    remoteRefContentMap = new Map(),
+    remoteRefSet = new Set(),
+    remoteRefResolutionPromises = [];
+
+  remoteRefResolver && traverseUtility(currentNode).forEach(function (property) {
+    if (property) {
+      let hasReferenceTypeKey;
+
+      hasReferenceTypeKey = Object.keys(property)
+        .find(
+          (key) => {
+            const isExternal = isExtURLRef(property, key),
+              isReferenciable = isExternal;
+
+            return isReferenciable;
+          }
+        );
+
+      if (hasReferenceTypeKey) {
+        const tempRef = calculatePath(parentFilename, property.$ref),
+          isRefEncountered = remoteRefSet.has(tempRef);
+
+        if (isRefEncountered) {
+          return;
+        }
+
+        remoteRefResolutionPromises.push(
+          new Promise(async (resolveInner) => {
+
+            /**
+             * Converts contents received from remoteRefResolver into stringified JSON
+             * @param {string | object} content - contents from remoteRefResolver
+             * @returns Stringified JSON contents
+             */
+            function convertToJSONString (content) {
+              if (typeof content === 'object') {
+                return JSON.stringify(content);
+              }
+
+              const parsedFile = parseFile(content);
+
+              return JSON.stringify(parsedFile.oasObject);
+            }
+
+            let contentFromRemote = await remoteRefResolver(property.$ref),
```

Comment:
We should have some handling in cases where this may fail and throw error.

we can either ignore and keep the field as is or mention it missing in bundled content somehow. But not handling it at all can cause issues.

Reply:
Handled err scenario